### ISSUE TITLE
[SelectionDAGISel] Mysteriously dropped chain on strict FP node.

### ISSUE THREAD

**Issue by JonPsson** [llvm/llvm-project#54617]
[testcase.tar.gz](https://github.com/llvm/llvm-project/files/8372375/testcase.tar.gz)

llc -mtriple=s390x-linux-gnu -mcpu=z15  ./tc_fpc.ll -o out.s -debug-only=isel

I started looking into the missing modelling of the Floating Point Control register on Calls (inspired by https://reviews.llvm.org/D111433, for PowerPC), and found to my surprise a particular opcode where the generated pattern matching does not preserve the chaining on the strict FP node.

More specifically, the first function in the attached test case makes two calls to @llvm.experimental.constrained.rint.v2f64(), separated by a call. Even without the (currently) missing FPC dependency on that call the chain dependencies should cause two separate VFIDB instructions ("VECTOR LOAD FP INTEGER") to be emitted, I believe. As seen in the debug output however, the chain is simply dropped:

```
ISEL: Starting selection on root node: t14: v2f64,ch = strict_frint nofpexcept t13, t2
ISEL: Starting pattern match
  Initial Opcode index to 42631
  TypeSwitch[v2f64] from 42633 to 42744
  Created node: t25: v2f64 = VFIDB nofpexcept t2, TargetConstant:i32<0>, TargetConstant:i32<0>
ISEL: Match complete!
```
The emitted MBB therefore has only one (reused) VFIDB, which is wrong as a called is crossed which may have changed the FPC setting.

If the type is changed from v2f64 to double, the chain is *not* dropped:

```
ISEL: Starting selection on root node: t12: f64,ch = strict_frint nofpexcept t11, t2
ISEL: Starting pattern match
  Initial Opcode index to 42631
  TypeSwitch[f64] from 42633 to 42636
  Morphed node: t12: f64,ch = WFIDB nofpexcept t2, TargetConstant:i32<0>, TargetConstant:i32<0>, t11
ISEL: Match complete!
```
The resulting MBB has one WFIDB before the call, and one after, which seems correct.

The patterns for VFIDB and WFIDB are both defined by the same VectorRounding multiclass, so it is strange that there is this difference. It is also worrisome: how can this happen without any assertion triggering? Is there any case where a chain is actually correct to drop, or could we add an assertion for this?

I see a comment in PPCInstrVSX.td "Rounding without exceptions (nearbyint). Due to strange tblgen behaviour, these need to be defined after the any_frint versions so ISEL will correctly add the chain to the strict versions.". Maybe this is related...?

And I see some 
`"OPC_MorphNodeTo1, TARGET_VAL(SystemZ::WFIDB), 0|OPFL_Chain" `
in SystemZGenDAGISel.inc, but only 
`"OPC_MorphNodeTo1, TARGET_VAL(SystemZ::VFIDB), 0"`
(no OPFL_Chain) for VFIDB.

(The following two functions in the test case are very similar except strict_fp_to_sint is used instead, which seem to be correctly handled).

**Comment by nemanjai** [comment]
Yes, the comment you pointed to around line 2800 of `llvm/lib/Target/PowerPC/PPCInstrVSX.td` comes from me hitting the same issue.

What appears to happen is that as table gen is processing the patterns, it determines whether the pattern should preserve the chain. If the first pattern is for matching a non-strict node, the chain isn't preserved. Hence we had to move the definition down. This is of course, just my speculation.

**Comment by topperc** [comment]
> Yes, the comment you pointed to around line 2800 of `llvm/lib/Target/PowerPC/PPCInstrVSX.td` comes from me hitting the same issue.
> 
> What appears to happen is that as table gen is processing the patterns, it determines whether the pattern should preserve the chain. If the first pattern is for matching a non-strict node, the chain isn't preserved. Hence we had to move the definition down. This is of course, just my speculation.

It's not based on pattern order. It's based on the pattern in the instruction definition. For that pattern this code in CodeGenDAGPatterns.cpp runs and sets `hasChain_Inferred` to true. If the pattern references a strict fp node we'll set  `InstInfo.hasChain` to true otherwise false.

```
  if (PatDef->isSubClassOf("Instruction")) {                                     
    InstInfo.isBitcast |= PatInfo.isBitcast;                                     
    InstInfo.hasChain |= PatInfo.hasChain;                                       
    InstInfo.hasChain_Inferred = true;                                           
  }   
```

later this code in DAGISelMatcherGen.cpp runs

```
    // For some instructions, we were able to infer from the pattern whether     
    // they should have a chain.  Otherwise, attach the chain to the root.       
    //                                                                           
    // FIXME2: This is extremely dubious for several reasons, not the least of   
    // which it gives special status to instructions with patterns that Pat<>    
    // nodes can't duplicate.
   if (II.hasChain_Inferred)                                                    
      NodeHasChain = II.hasChain;                                                
    else                                                                         
      NodeHasChain = isRoot;                                                     
    // Instructions which load and store from memory should have a chain,        
    // regardless of whether they happen to have a pattern saying so.            
    if (II.hasCtrlDep || II.mayLoad || II.mayStore || II.canFoldAsLoad ||        
        II.hasSideEffects)                                                       
      NodeHasChain = true;     
```

Because `hasChain_Inferred` was set we take the first `if` and set `NodeHasChain` to false. The second `if` is missing `mayRaiseFPException` so we don't override it to true like we do for load/store/etc.

If the instruction pattern contains `any_*` we infer `II.hasChain` to true so this problem doesn't usually appear.

**Comment by topperc** [comment]
Poking around a little there are definitely similar issues on RISCV and X86. I think we need to add mayRaiseFPException to the code in DAGISelMatcherGen.cpp. I'm wondering if we should only do it if the pattern tree contains a strict FP node. I don't think tablegen knows which SDNodes are strictfp so we would probably need a new SDNode property.

**Comment by topperc** [comment]
Attempted patch here https://reviews.llvm.org/D127254
